src/box/applier.cc

```diff
@@ -1290,6 +1346,682 @@ applier_on_rollback(struct trigger *trigger, void *event)
 	return 0;
 }
 
+/** The underlying thread behind a number of appliers. */
+struct applier_thread {
+	/** A link in allocated threads list. */
+	struct rlist in_list;
+	struct cord cord;
+	/** The single thread endpoint. */
+	struct cbus_endpoint endpoint;
+	/** A pre-allocated message emitted by the thread once it's exiting. */
+	struct cmsg join_msg;
+	/** Count of appliers attached to this thread. */
+	int n_appliers;
+	/** Whether the thread has no appliers and has to be joined. */
+	bool is_exiting;
+};
+
+/**
+ * Notify tx that the applier thread has detached its last applier and has to
+ * be joined.
+ */
+static void
+applier_thread_set_joinable(struct cmsg *msg)
+{
+	struct applier_thread *thread = container_of(msg, struct applier_thread,
+						     join_msg);
+	thread->is_exiting = true;
+}
+
+static const struct cmsg_hop join_msg_route[] = {
+	{applier_thread_set_joinable, NULL},
+};
+
+struct applier_detach_msg {
+	struct cmsg base;
+	struct diag diag;
+};
+
+/**
+ * Propagate the exception causing the applier death to tx from applier
+ * thread.
+ */
+static void
+applier_detach_thread(struct cmsg *base)
+{
+	struct applier_detach_msg *msg = (struct applier_detach_msg *)base;
+	assert(!diag_is_empty(&msg->diag));
+	diag_move(&msg->diag, &fiber()->diag);
+	diag_raise();
+}
+
+static const struct cmsg_hop detach_msg_route[] = {
+	{applier_detach_thread, NULL},
+};
+
+struct applier_msg {
+	struct cmsg base;
+	struct applier *applier;
+	/**
+	 * The last known confirmed vclock to be used by applier ack writer
+	 * fiber.
+	 */
+	struct vclock ack_vclock;
+	/** Whether the message has an updated ack_vclcok value. */
+	bool has_ack;
+	/** A list of read up transactions to be processed in tx thread. */
+	struct stailq txs;
+	/** This message's personal buffer, holding input rows. */
+	struct ibuf ibuf;
+	/** This message's region used to allocate auxiliary structures. */
+	struct region region;
+	int txn_cnt;
+};
+
+/* Applier thread related data. */
+struct applier_thread_data {
+	/** A pointer to the thread handling this applier's data stream. */
+	struct applier_thread *thread;
+	/** Applier's personal endpoint in the tx thread. */
+	struct cbus_endpoint endpoint;
```

Comment:
Why not reuse "tx" endpoint?

Reply:
Because I want the applier fiber to process all the corresponding messages. Not just some random fiber from the tx fiber pool.

This way it's easier to keep applier reconnect logic (when applier_f retries connection each time it catches an exception).

This is not a big deal, and could be done with "tx" fiber pool, but I don't see any downsides to the endpoint-per-applier approach. Do you?

Reply:
Discussed f2f. Agreed that adding extra tx endpoints may affect tx performance. Decided to use the `tx_prio` endpoint for pushing applier messages to tx.